### PR TITLE
PR to develop

### PR DESCRIPTION
**Pull request by cfournel**

### CODE REVIEW

File: src/.vscode/c_cpp_properties.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
+    "configurations": [
+        {
+            "name": "Linux",
+            "includePath": [
```
Comment (cujomalainey) [inline src/.vscode/c_cpp_properties.json:5]: This file should not be part of the PR as it is local to your machine

File: examples/FecBikeTrainerDisplay/fe-c_native.ino [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+#include <Arduino.h>
+#include "ANT.h"
+#include "ANTPLUS.h"
+#include "bluefruit.h"
+#define BAUD_RATE 115200
+#define CHANNEL_0 0
+#define TOTAL_CHANNELS 1
+#define ENCRYPTED_CHANNELS 0
+
+const uint8_t NETWORK_KEY[] = {0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08}; // get this from thisisant.com
```
Comment (cujomalainey) [inline examples/FecBikeTrainerDisplay/fe-c_native.ino:18]: In one of your commits you have this number, you can squash the two commits so the history is removed or I can do it.

File: examples/FecBikeTrainerDisplay/fe-c_native.ino [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+ArduinoNativeAntWithCallbacks ant;
```
Comment (cujomalainey) [inline examples/FecBikeTrainerDisplay/fe-c_native.ino:20]: Unfortunately this has to be serial otherwise I can't add it as part of the CI.

File: examples/FecBikeTrainerDisplay/fe-c_native.ino [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+     * 
+     * Just a quick test to switch target power 
+     * every **interval**
+     * 
+     */
+    if(currentMillis - previousMillis > interval) 
+    {
+      // save the last time you blinked the LED 
+      previousMillis = currentMillis;   
+      //Serial.println("Envoi de la puissance vise");
```
Comment (cujomalainey) [inline examples/FecBikeTrainerDisplay/fe-c_native.ino:107]: Is this debug code?

**Review by cujomalainey — CHANGES_REQUESTED** [review]
Overall it looks pretty good from an initial look. I just did an initial pass, I will try and take a look at it with the spec in hand tonight to verify the byte patterns. Are you familiar enough with git to squash to remove that number or do you want me to do it after submission?

File: src/Profiles/Fec/ANTPLUS_FecDataPages.h [hunk 5]
```diff
@@ -2,5 +2,14 @@
 #define ANTPLUS_PROFILEFECDATAPAGES_h
 
 #include <Profiles/Fec/DataPages/ANTPLUS_FecSpecificTrainerData.h>
+#include <Profiles/Fec/DataPages/ANTPLUS_FecManufacturerInformation.h>
+#include <Profiles/Fec/DataPages/ANTPLUS_FecProductInformation.h>
```
Comment (cujomalainey) [inline src/Profiles/Fec/ANTPLUS_FecDataPages.h:6]: According to the spec, manufacturers and product information are common datapages, we should use the existing types for that in the common datapages area

File: src/Profiles/Fec/ANTPLUS_FecPrivateDefines.h [hunk 6]
```diff
@@ -10,6 +10,16 @@
 // 30 / 2.5 = 12
 #define ANTPLUS_FEC_SEARCHTIMEOUT 12
 
+/* DataPages number */
+#define ANTPLUS_FEC_MANUFACTURERINFORMATION_NUMBER 80
+#define ANTPLUS_FEC_PRODUCT_INFORMATION_NUMBER 81
 #define ANTPLUS_FEC_DATAPAGE_TRAINER_DATA_NUMBER 25
+#define ANTPLUS_FEC_GENERAL_INFORMATION_NUMBER 16
```
Comment (cujomalainey) [inline src/Profiles/Fec/ANTPLUS_FecPrivateDefines.h:17]: GENERAL_FE_DATA to match spec

File: src/Profiles/Fec/ANTPLUS_FecPrivateDefines.h [hunk 7]
```diff
@@ -10,6 +10,16 @@
...
+/* DataPages number */
+#define ANTPLUS_FEC_MANUFACTURERINFORMATION_NUMBER 80
+#define ANTPLUS_FEC_PRODUCT_INFORMATION_NUMBER 81
 #define ANTPLUS_FEC_DATAPAGE_TRAINER_DATA_NUMBER 25
+#define ANTPLUS_FEC_GENERAL_INFORMATION_NUMBER 16
+#define ANTPLUS_FEC_GENERAL_SETTINGS_NUMBER 17
+#define ANTPLUS_FEC_BASIC_RESISTANCE_NUMBER 48
+#define ANTPLUS_FEC_TARGET_POWER_NUMBER 49
+#define ANTPLUS_FEC_TRACK_RESISTANCE_NUMBER 51
+#define ANTPLUS_FEC_CAPABILITIES_INFORMATION_NUMBER 54
```
Comment (cujomalainey) [inline src/Profiles/Fec/ANTPLUS_FecPrivateDefines.h:22]: FE_CAPABILITIES

File: src/Profiles/Fec/ANTPLUS_FecPrivateDefines.h [hunk 8]
```diff
@@ -10,6 +10,16 @@
...
+#define ANTPLUS_FEC_USER_INFORMATION_NUMBER 55
```
Comment (cujomalainey) [inline src/Profiles/Fec/ANTPLUS_FecPrivateDefines.h:23]: USER_CONFIGURATION

File: src/Profiles/Fec/DataPages/ANTPLUS_FecBaseMainDataPage.cpp [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    return this->get8BitValue(DATAPAGE_BYTE, DATAPAGE_MASK);
+}
+
+template<class T>
+uint8_t FecCoreMainDataPage<T>::getPageChangeToggle() {
+    return this->get8BitValue(TOGGLE_BYTE, TOGGLE_MASK, TOGGLE_SHIFT);
+}
+
+template<class T>
+uint16_t FecCoreMainDataPage<T>::getHeartBeatEventTime() {
```
Comment (cujomalainey) [inline src/Profiles/Fec/DataPages/ANTPLUS_FecBaseMainDataPage.cpp:29]: I should really document this, but the "core" datapages are meant to the base and handle any common functionality. For example. HR uses custom datapages where all the fields are the same except 3 bytes so a bunch of the fields are in there. I don't think FEC has these so these should probably be removed. The type can stay if you see value in using it to get the page number.

File: src/Profiles/Fec/DataPages/ANTPLUS_FecBasicResistanceDataPage.cpp [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+#include <Profiles/Fec/ANTPLUS_FecPrivateDefines.h>
+
+#define POWER_BYTE 7
+
+template<class T>
+FecBaseBasicResistanceDataPage<T>::FecBaseBasicResistanceDataPage() :
+    CoreDataPage<T>() {}
+
+template<class T>
+uint8_t FecBaseBasicResistanceDataPage<T>::getPower() {
```
Comment (cujomalainey) [inline src/Profiles/Fec/DataPages/ANTPLUS_FecBasicResistanceDataPage.cpp:11]: where does "power" come from? This field is labeled as "total resistance" in my doc

File: src/Profiles/Fec/DataPages/ANTPLUS_FecGeneralMainDataPage.cpp [hunk 11]
```diff
@@ -16,50 +16,61 @@
...
 template<class T>
-uint16_t FECGeneralMainDataPage<T>::getSpeed() {
+/*
+
+in meter / s 
+
+*/
+uint16_t FecBaseGeneralMainDataPage<T>::getSpeed(bool km) {
+    if ( km == true ){
+        return 3.6 * (this->get16BitValue(SPEED_LSB_BYTE, SPEED_MSB_BYTE)) /1000;
```
Comment (cujomalainey) [inline src/Profiles/Fec/DataPages/ANTPLUS_FecGeneralMainDataPage.cpp:62]: you can do this in oneline like this
`return  (kph ? 3.6 : 1.0) * (the rest of the function call)`

also 1m/s is 3.6 kph, so why the /1000? but also I like to avoid conversions in my lib, and just provide the data in the expected format just for safety given these values could be converted a number of ways by end users.

File: src/Profiles/Fec/DataPages/ANTPLUS_FecGeneralMainDataPage.h [hunk 12]
```diff
@@ -1,22 +1,31 @@
...
-#endif // ANTPLUS_FECCOREMAINDATAPAGE_h
+class FecGeneralMainDataPage : public FecBaseMainDataPage, public FecBaseGeneralMainDataPage<BroadcastData> {
+public:
+    explicit FecGeneralMainDataPage(AntRxDataResponse& dp);
+};
+
+class FecGeneralMainDataMsg : public FecBaseMainDataPageMsg, public FecBaseGeneralMainDataPage<BroadcastDataMsg> {
+public:
+    FecGeneralMainDataMsg();
+  //  void setSoftwareRevision(uint16_t SoftwareRevision);
```
Comment (cujomalainey) [inline src/Profiles/Fec/DataPages/ANTPLUS_FecGeneralMainDataPage.h:27]: mind replacing this comment with `TODO`?

File: src/Profiles/Fec/DataPages/ANTPLUS_FecGeneralSettingsDataPage.cpp [hunk 13]
```diff
@@ -0,0 +1,60 @@
+#include <Profiles/Fec/DataPages/ANTPLUS_FecGeneralSettingsDataPage.h>
+#include <Profiles/Fec/ANTPLUS_FecPrivateDefines.h>
+#include <ANTPLUS_PrivateDefines.h>
```
Comment (cujomalainey) [inline src/Profiles/Fec/DataPages/ANTPLUS_FecGeneralSettingsDataPage.cpp:3]: i don't see the private define used here, where is it?

File: src/Profiles/Fec/DataPages/ANTPLUS_FecGeneralSettingsDataPage.h [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+
+class FecGeneralSettingsDataPage : public FecBaseMainDataPage, public FecBaseGeneralSettingsDataPage<BroadcastData> {
+public:
+    explicit FecGeneralSettingsDataPage(AntRxDataResponse& dp);
+};
+
+class FecGeneralSettingsDataMsg : public FecBaseMainDataPageMsg, public FecBaseGeneralSettingsDataPage<BroadcastDataMsg> {
+public:
+    FecGeneralSettingsDataMsg();
+    void setResistanceLevel(uint8_t ResistanceLevel);
```
Comment (cujomalainey) [inline src/Profiles/Fec/DataPages/ANTPLUS_FecGeneralSettingsDataPage.h:25]: please add TODO

File: src/Profiles/Fec/DataPages/ANTPLUS_FecGeneralSettingsDataPage.cpp [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+FecGeneralSettingsDataPage::FecGeneralSettingsDataPage(AntRxDataResponse& dp) :
+    FecBaseMainDataPage(dp),
+    FecBaseGeneralSettingsDataPage<BroadcastData>() {}
+
+FecGeneralSettingsDataMsg::FecGeneralSettingsDataMsg() :
+    FecBaseMainDataPageMsg(ANTPLUS_FEC_GENERAL_SETTINGS_NUMBER),
+    FecBaseGeneralSettingsDataPage<BroadcastDataMsg>() {}
+
+void FecGeneralSettingsDataMsg::setResistanceLevel(uint8_t ResistanceLevel) {
+    set8BitValue(ResistanceLevel, RESISTANTCE_LEVEL_BYTE);
```
Comment (cujomalainey) [inline src/Profiles/Fec/DataPages/ANTPLUS_FecGeneralSettingsDataPage.cpp:59]: please add TODO

File: src/Profiles/Fec/DataPages/ANTPLUS_FecManufacturerInformation.cpp [hunk 16]
```diff
@@ -0,0 +1,54 @@
+#include <Profiles/Fec/DataPages/ANTPLUS_FecManufacturerInformation.h>
```
Comment (cujomalainey) [inline src/Profiles/Fec/DataPages/ANTPLUS_FecManufacturerInformation.cpp:1]: this should be substitutable by the stuff in common datapages

File: src/Profiles/Fec/DataPages/ANTPLUS_FecProductInformation.cpp [hunk 17]
```diff
@@ -0,0 +1,31 @@
+#include <Profiles/Fec/DataPages/ANTPLUS_FecProductInformation.h>
```
Comment (cujomalainey) [inline src/Profiles/Fec/DataPages/ANTPLUS_FecProductInformation.cpp:1]: this should be substitutable by the stuff in common datapages

File: src/Profiles/Fec/DataPages/ANTPLUS_FecTrackResistanceDataPage.cpp [hunk 18]
```diff
@@ -0,0 +1,31 @@
+#include <Profiles/Fec/DataPages/ANTPLUS_FecTrackResistanceDataPage.h>
+#include <Profiles/Fec/ANTPLUS_FecPrivateDefines.h>
+#include <ANTPLUS_PrivateDefines.h>
+
+// TODO this is a mixture of datapages
```
Comment (cujomalainey) [inline src/Profiles/Fec/DataPages/ANTPLUS_FecTrackResistanceDataPage.cpp:5]: what do you mean by this comment? This looks like one page, although yes there is a TODO as last field is missing.

File: src/Profiles/Fec/DataPages/ANTPLUS_FecUserInformationDataPage.cpp [hunk 19]
```diff
@@ -0,0 +1,42 @@
+#include <Profiles/Fec/DataPages/ANTPLUS_FecUserInformationDataPage.h>
```
Comment (cujomalainey) [inline src/Profiles/Fec/DataPages/ANTPLUS_FecUserInformationDataPage.cpp:1]: is this supposed to be UserConfiguration? I dont see a user information page

File: src/Profiles/Fec/DataPages/ANTPLUS_FecUserInformationDataPage.cpp [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+#define BICYCLE_WEIGHT_LSB_BYTE 4
+#define BICYCLE_WEIGHT_MSB_BYTE 5
+
+template<class T>
+FecBaseUserInformationDatapage<T>::FecBaseUserInformationDatapage() :
+    CoreDataPage<T>() {}
+
+template<class T>
+uint16_t FecBaseUserInformationDatapage<T>::getUserWeight() {
+    return this->get16BitValue(USER_WEIGHT_LSB_BYTE, USER_WEIGHT_MSB_BYTE);
```
Comment (cujomalainey) [inline src/Profiles/Fec/DataPages/ANTPLUS_FecUserInformationDataPage.cpp:16]: please add a TODO so I know all the fields were not added

**Review by cujomalainey — CHANGES_REQUESTED** [review]
Some minor issues but overall looks really good, I will hunt down why CI doesn't think it needs to run. Also you can check out Lev or Bicycle Power for examples on how to use common datapages. Thanks again for the contribution

**Comment by cfournel** [comment]
Hi Curtis,


I don't know how to squash but tell me how and i will do it



Charles-antoine FOURNEL

Le jeu. 13 janv. 2022 à 23:51, Curtis Malainey ***@***.***> a
écrit :

**Comment by cfournel** [comment]
Also,

From datapage 51 sheet, you can set up grade simulation from -200 to 200% i
only write getter and setter and did not propose functions to set the grade
with ease

For example if you want to simulate a grade slope of 5 % you have to send
an uint_16 of 20500 ( unit is 0.01 ) because of the -200 offset specified
so its like 20500 * 0.01 = 205 - 200 = 5 which is thé targeted grade.

In your antplus philosophy , should this bé implemented or its up to
everyone to handle it ?



Charles-antoine FOURNEL

**Comment by cujomalainey** [comment]
Hi Charles,

If you don't know how to squash then i should probably do it since it is a destructive action to git history so it can be dangerous if done incorrectly and cause you to lose work.

As for the philosophy I usually avoid those helpers as everyone may have a different preference for scale, but the benefit is you can write your own class in your app and extend it yourself and use it in place of the original message for both TX and RX. For TX you just use your type, for RX you just load the data pointer from the original message similar to how its done in the display from the original message then you can use the extensions.

**Comment by cujomalainey** [comment]
Hello Charles

Do you have time to address these comments or should I merge and squash them and address them later?

**Comment by cujomalainey** [comment]
merging and going to handle rest of the cleanup in tree

**Comment by cujomalainey** [comment]
update is squashed into original commit now as 3d695b9094d51848754c377baae93b85bdc506a6 so numbers have been removed from history

File: src/Profiles/Fec/DataPages/ANTPLUS_FecCapabilitiesInformationDataPage.cpp [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+
+template class FecBaseCapabilitiesInformationDatapage<BroadcastData>;
+template class FecBaseCapabilitiesInformationDatapage<BroadcastDataMsg>;
+
+FecCapabilitiesInformationDatapage::FecCapabilitiesInformationDatapage(AntRxDataResponse& dp) :
+    FecBaseMainDataPage(dp),
+    FecBaseCapabilitiesInformationDatapage<BroadcastData>() {}
+
+FecCapabilitiesInformationMsg::FecCapabilitiesInformationMsg() :
+    FecBaseMainDataPageMsg(ANTPLUS_FEC_USER_INFORMATION_NUMBER),
```
Comment (cujomalainey) [inline src/Profiles/Fec/DataPages/ANTPLUS_FecCapabilitiesInformationDataPage.cpp:30]: wong dp number here
